lib/route.js

```diff
@@ -222,23 +222,23 @@ function buildRouting (options) {
       // not called for every single route. Creating a new one for every route
       // is going to be very expensive.
       if (opts.schema) {
-        if (this[kSchemaCompiler] == null && this[kSchemaResolver]) {
-          done(new FST_ERR_SCH_MISSING_COMPILER(opts.method, url))
-          return
-        }
-
-        try {
-          if (opts.schemaCompiler == null && this[kSchemaCompiler] == null) {
-            const externalSchemas = this[kSchemas].getJsonSchemas({ onlyAbsoluteUri: true })
-            this.setSchemaCompiler(buildSchemaCompiler(externalSchemas, this[kOptions].ajv, schemaCache))
+        this.preReady(() => {
+          if (this[kSchemaCompiler] == null && this[kSchemaResolver]) {
+            throw new FST_ERR_SCH_MISSING_COMPILER(opts.method, url)
           }
 
-          buildSchema(context, opts.schemaCompiler || this[kSchemaCompiler], this[kSchemas], this[kSchemaResolver])
-        } catch (error) {
-          // bubble up the FastifyError instance
-          done(error.code ? error : new FST_ERR_SCH_BUILD(opts.method, url, error.message))
-          return
-        }
+          try {
+            if (opts.schemaCompiler == null && this[kSchemaCompiler] == null) {
+              const externalSchemas = this[kSchemas].getJsonSchemas({ onlyAbsoluteUri: true })
+              this.setSchemaCompiler(buildSchemaCompiler(externalSchemas, this[kOptions].ajv, schemaCache))
+            }
+
+            buildSchema(context, opts.schemaCompiler || this[kSchemaCompiler], this[kSchemas], this[kSchemaResolver])
+          } catch (error) {
+            // bubble up the FastifyError instance
+            throw error.code ? error : new FST_ERR_SCH_BUILD(opts.method, url, error.message)
+          }
+        })
```

Comment:
THere is no need to change how errors are handled. just move the if-block down in the `preReady` event handler. See https://github.com/fastify/fastify/pull/2077/files#diff-f34e409694d3ba334dabbea190f64c1dR264

Reply:
Oh, I have overlooked that part!!